### ISSUE TITLE
"Journaling" cache

### ISSUE THREAD

**Issue by daggaz** [archiecobbs/s3backer#87]
In the case of a unclean shutdown (e.g. server power loss, software crash, whatever) any blocks in the  cache in the `DIRTY`, `WRITING` or `WRITING2` states will be discarded.

If a block in the cache is in once of these states then the cache data must be the correct version.

Can we record the in the on-disk cache in the `dir_entry` structure that the block needs to be written. Then on startup with an existing cache file, the blocks marked with this flag will be added to the dirty list (in `block_cache_dcache_load()`).

This all assumes that the filesystem has not been mounted elsewhere. We can record in the cache file the same data that is written to the `s3backer-mounted` file and this would indicate the cache file can be used for recovery (and not prevent mounting).

**Comment by Gazzonyx** [comment]
Sounds reasonable. But, can the ordering of the blocks written be guaranteed so you’re certain to write your “journal” entry before the data block? What I mean is, doesn’t this become a race between the actual file system and journal of the cache file vs. the “journaling” we’d be doing and the file system and journal in S3? Even more simply, how many states can the two real file systems (cache file backing and file system on top of S3B - one local, one network) and their journals be in at a random moment after we write our intent that the block is dirty and then write to S3? If I’m not mistaken, this ordering will matter?

**Comment by daggaz** [comment]
The cache journal and cache data are written syncronously in the fuse callback. If the crash happens in that callback, then the filesystem on top of S3B will deal with that (assuming it is itself a journalling filesystem).

After the synchronous callback then a journalling filesystem on top will assume that the write was successful. The problem occurs if the system crashes after the synchronous cache write, but before the write to S3 is complete (i.e. the block is dirty). If that dirty block is discarded it violates the assumption made by the file system on top that the data has really been written.

**Comment by archiecobbs** [comment]
I think there are a couple of distinct issues here.

1. The "Hey! You said you durably persisted that block but you lied!" problem. In other words, the filesystem asks the disk to write a block _and then confirm that the block was durably persisted_, the disk confirms it, and yet if there is a crash at the wrong moment, the block was not actually durably persisted.

2. The blocks being written out of order problem. The filesystem says to durably persist blocks 1, 2, 3, 4, 5, but because of re-ordering, if there is a crash at the wrong moment, it will appear as if blocks 2 and 4 were persisted, but not 1, 3, or 5. This would only be a problem if the filesystem invoked some kind of write barrier at the block device layer, because (I think) block devices don't guarantee any ordering by default.

Note both of these issues can only occur if there is a crash (unclean shutdown) of some kind.

Problem 1 is only a problem if you're running applications that use `fsync(2)` or `fdatasync(3)` to guarantee durability, e.g., a transactional database, email server (store-and-forward), etc.

Problem 2 is potentially much more serious because it can lead to corruption of the filesystem's data structures. Some filesystems protect against this (though I don't know the details). For example, if a journalling filesystem includes a checksum in each journal segment, then the first discontinuity in the journal will be noticed and the journal truncated at that point; this avoids problem 2 but not problem 1.

What happens when a filesystem asks the "block device" to confirm a block has been durably persisted, and the "block device" is actually a loopback mount? I have no idea. One might hope that the kernel returns a conservative answer in this situation.

It would be interesting to ask some kernel filesystem experts what really happens in the details.

**Comment by daggaz** [comment]
Here's a quick PR for what I'm thinking off.

A good analogy here is a battery backed cache on a hardware raid controller. That's all we're trying to emulate.

**Comment by daggaz** [comment]
I'm working on further changes to detect that the cache file and the s3backer-mounted file in s3 actually match up (meaning you don't have to --reset-mounted-flag or --force if you have a recoverable cache).

I tried this with XFS on top and killed the s3backer process while the dirty ratio > 0.

On remounting s3backer and the XFS filesystem, XFS was able to recover from its journal just fine and you could see all the dirty blocks in the cache being written to s3.

**Comment by daggaz** [comment]
On the disk-layout question, what would be the best option? I see three options if a cache file with an old signature is found:

1. Fail with helpful error message
   This means that an upgrade and remount would break without manual intervention
1. Throw away old cache data and start new file (silently, or with warning)
   This provides a seamless upgrade. Since existing cache data is currently throw-away, this should not be a problem.
1. Rewrite the cache on disk
    Starting from the end, extend the file, shift all the data, rewrite header and directory
    This is seamless, but could take a significant time to mount on startup for a large existing cache

I think we prefer option 2.

**Comment by archiecobbs** [comment]
I like option 1 better. It's effectively the same as option 2 except that it requires explicit confirmation by the user, which is more conservative. We don't really know whether the user is willing to throw away all their currently cached data just to get the new feature.

So you'd get something like this...

```
$ s3backer ... --blockCacheFile cachefile --blockCacheFlushWritableOnStartup
s3backer: error: cache file `cachefile' is not compatible with `--blockCacheFlushWritableOnStartup'
$ rm cachefile
$ s3backer ... --blockCacheFile cachefile --blockCacheFlushWritableOnStartup
... no errors...
```

**Comment by daggaz** [comment]
I don't think we want to code for support of two different cache file formats (especially since everything is based off `sizeof()` of the `file_header` and `dir_entry` structs and that would have to replaced with piles of `if` statements).

The change to the cache file format is required regardless of the option being set or not so taking option 1 means that the only recourse is to manually delete the cache file anyway.  It just means that upgrading is not seamless.

I also don't like the idea that when someone does an upgrade, their filesystem won't just mount.

**Comment by archiecobbs** [comment]
> I also don't like the idea that when someone does an upgrade, their filesystem won't just mount.

Agreed, except that some users may prefer that to the alternative - i.e., all of their cached data being silently blown away, resulting in a burst of network activity, S3 access charges, and a massive slowdown as the cache reloads. 

I don't feel confident assuming what people might want here, especially since dollars may be involved.

**Comment by daggaz** [comment]
So what's the answer then?

We either:
* accept the error on upgrade
* go down the support multiple formats route (crazy, and maintenance nightmare)
* attempt the in-place cache file upgrade

**Comment by archiecobbs** [comment]
I'm afraid the multiple formats route currently gets my vote... but to help out I've added (untested) compatibility code in 9ea1f31.

Still missing is the implementation of `--blockCacheFlushWritableOnStartup`, which if this flag is not set, should erase the dirty cache file entries. This will take some adjustment to the "visitor".

**Comment by leo60228** [comment]
On Problem 2, I believe that this presents a major possibility of corruption for btrfs. According to [this FAQ](https://btrfs.wiki.kernel.org/index.php/FAQ#What_does_.22parent_transid_verify_failed.22_mean.3F), barriers are used to ensure that the superblock is written after other blocks. According to [an earlier FAQ](https://btrfs.wiki.kernel.org/index.php/FAQ#I_see_a_warning_in_dmesg_about_barriers_being_disabled_when_mounting_my_filesystem._What_does_that_mean.3F), this occurring **will** cause full filesystem corruption.

**Comment by daggaz** [comment]
We've already got a new visitor implementation. I'll push it when I get to work.

**Comment by archiecobbs** [comment]
The problem with Problem 2 is that the only way for `s3backer` to guarantee in-order writes is to write one block at a time, which is really slow.

Now if you have a huge local cache, and write bursts smaller than your cache size, this might not be a problem.

But because there's no FUSE operation that says "write barrier now", we would have to conservatively assume that there is an implicit write barrier after every block written.

**Comment by daggaz** [comment]
Just to confirm the requirements here:
* existing cache files should work (backward compatible)
* new cache files are created in the new format
* new cache files can be mounted with the new option, and this will recover any dirty data

One question, in your commit in [`dcache.c:348`](https://github.com/archiecobbs/s3backer/blob/9ea1f3162460d328401cb0c9258825a3ce636c2a/dcache.c#L348) you explicitly erase the block in the directory (preserving the old behaviour exactly). However, if the entry was always preserved in the directory, then a new-format cache file would be recoverable even if the new option was not specified at the time of the crash (you'd only need to specify it after the crash). If the option is not specified, dirty entries can be thrown away at startup (in the visitor).

Essentially, it'd be nice if you only need to specify the flag to perform the recovery. There is no downside (I can see) to recording the dirty entries regardless, and handling them on startup depending on the option set.

**Comment by daggaz** [comment]
We already have a working version of the code that does this, but its not yet ready to be pushed.

**Comment by archiecobbs** [comment]
> Just to confirm the requirements here:
> * existing cache files should work (backward compatible)
> * new cache files are created in the new format
> * new cache files can be mounted with the new option, and this will recover any dirty data

All correct.

> One question, in your commit in [`dcache.c:348`](https://github.com/archiecobbs/s3backer/blob/9ea1f3162460d328401cb0c9258825a3ce636c2a/dcache.c#L348) you explicitly erase the block in the directory (preserving the old behaviour exactly). However, if the entry was always preserved in the directory, then a new-format cache file would be recoverable even if the new option was not specified at the time of the crash (you'd only need to specify it after the crash). 

The behavior in this `if` statement is a function of the file format, not whether the command-line flag is present. So the erasure will never happen on a new-format file (see [`dcache.c:242`](https://github.com/archiecobbs/s3backer/blob/9ea1f3162460d328401cb0c9258825a3ce636c2a/dcache.c#L242)).

But there is indeed a stupid bug in that line, it should be:

```
    if (needs_write && (priv->flags & HDRFLG_NEW_DIRENTRY) == 0) {
```

Fixed in 15e5769b.

**Comment by daggaz** [comment]
Great, my mistake.

**Comment by daggaz** [comment]
That is I think a complete implementation. Still doing a bit of local testing.

@archiecobbs can you take a look and see what you think?

**Comment by archiecobbs** [comment]
I'll take a look. It may take a while to complete.

I like the idea of the setting a unique mount cookie, instead of a simple flag, for safety.

**Comment by daggaz** [comment]
Added some extra logging information when performing a cache recovery

**Comment by archiecobbs** [comment]
Current working code is in branch `daggaz-issue-87`.

I did a bunch of refactoring and merging and have not tested this much. But the diff is now much smaller vs. `master` than before. Please review.

**Comment by daggaz** [comment]
We've reviewed the diffs and think your changes are all sensible. We're going to do some testing.

I don't know your usual release process, but I suspect that CHANGES needs updated and the version number bumped at least.

**Comment by daggaz** [comment]
We tested it all and didn't find any problems.

**Comment by archiecobbs** [comment]
In `block_cache_write()`, you added a MD5 buffer `md5` which is passed to `s3b_dcache_record_block()` without being initialized.

I think the `md5` field is useless for dirty entries. It seems only there for the benefit of `block_cache_dcache_load()` at restart when reloading the cache from disk. But `block_cache_dcache_load()` ignores the MD5 when the block is dirty.

So this `md5` buffer can be removed, and the parameter can just be passed as null.

Do you concur? See 78a0398.

**Comment by daggaz** [comment]
Yes, I never liked that empty MD5, but, I'm not sure we should just pass a NULL pointer to memcpy and write random data from the stack to disk. Perhaps we should memset the entry to 0 instead?

**Comment by Gazzonyx** [comment]
I'm much more in favor of this solution (zeroing out the memory) even if
only because it makes debugging easier. In my humble opinion, it's cleaner
than the alternative.

**Comment by archiecobbs** [comment]
@daggaz , see subsequent commits; the first one had a null pointer bug and would have crashed.

**Comment by daggaz** [comment]
Yep, looks good to me.
